go/apps/api/routes/v2_keys_create_key/handler.go

```diff
@@ -358,112 +350,233 @@ func (h *Handler) Handle(ctx context.Context, s *zen.Session) error {
+					WorkspaceID:  auth.AuthorizedWorkspaceID,
+					Name:         requestedSlug,
+					Slug:         requestedSlug,
+					Description:  sql.NullString{String: fmt.Sprintf("Auto-created permission: %s", requestedSlug), Valid: true},
+					CreatedAtM:   now,
+				})
+
+				requestedPermissions = append(requestedPermissions, db.FindPermissionsBySlugsRow{
+					ID:   newPermID,
+					Slug: requestedSlug,
+				})
+			}
+
+			if len(permissionsToCreate) > 0 {
+				err = db.BulkInsert(ctx, tx,
+					"INSERT INTO permissions (id, workspace_id, name, slug, description, created_at_m) VALUES (?, ?, ?, ?, ?, ?)",
+					permissionsToCreate,
+				)
+
+				if err != nil {
+					return fault.Wrap(err,
+						fault.Code(codes.App.Internal.ServiceUnavailable.URN()),
+						fault.Internal("database error"),
+						fault.Public("Failed to create permissions."),
+					)
+				}
+			}
+
+			permissionsToInsert := []db.InsertKeyPermissionParams{}
+			for _, reqPerm := range requestedPermissions {
+				permissionsToInsert = append(permissionsToInsert, db.InsertKeyPermissionParams{
+					KeyID:        keyID,
+					PermissionID: reqPerm.ID,
+					WorkspaceID:  auth.AuthorizedWorkspaceID,
+					CreatedAt:    now,
+				})
+
+				auditLogs = append(auditLogs, auditlog.AuditLog{
+					WorkspaceID: auth.AuthorizedWorkspaceID,
+					Event:       auditlog.AuthConnectPermissionKeyEvent,
+					ActorType:   auditlog.RootKeyActor,
+					ActorID:     auth.Key.ID,
+					ActorName:   "root key",
+					ActorMeta:   map[string]any{},
+					Display:     fmt.Sprintf("Added permission %s to key %s", reqPerm.Slug, keyID),
+					RemoteIP:    s.Location(),
+					UserAgent:   s.UserAgent(),
+					Resources: []auditlog.AuditLogResource{
+						{
+							Type:        "key",
+							ID:          keyID,
+							Name:        insertKeyParams.Name.String,
+							DisplayName: insertKeyParams.Name.String,
+							Meta:        map[string]any{},
+						},
+						{
+							Type:        "permission",
+							ID:          reqPerm.ID,
+							Name:        reqPerm.Slug,
+							DisplayName: reqPerm.Slug,
+							Meta:        map[string]any{},
+						},
 					},
-				},
-			})
+				})
+			}
+
+			if len(permissionsToInsert) > 0 {
+				err = db.BulkInsert(ctx, tx,
+					"INSERT INTO keys_permissions (key_id, permission_id, workspace_id, created_at_m) VALUES (?, ?, ?, ?)",
+					permissionsToInsert,
+				)
+				if err != nil {
+					return fault.Wrap(err,
+						fault.Code(codes.App.Internal.ServiceUnavailable.URN()),
+						fault.Internal("database error"),
+						fault.Public("Failed to assign permissions."),
+					)
+				}
+			}
 		}
 
-		if len(rolesToInsert) > 0 {
-			err = db.BulkInsert(
-				ctx,
-				tx,
-				"INSERT INTO keys_roles (key_id, role_id, workspace_id, created_at_m) VALUES (?, ?, ?, ?)",
-				rolesToInsert,
-			)
+		// 12. Handle roles if provided - with auto-creation
+		if req.Roles != nil {
+			existingRoles, err := db.Query.FindRolesByNames(ctx, tx, db.FindRolesByNamesParams{
```

Comment:
Does upsetting roles really make sense though
A newly created role would have 0 permissions

I mean it would be weird not to autocreate if everything else is auto created but this feels a bit weird

Reply:
yeahhh not really